File: src/core/Sfx.ts
import type Phaser from 'phaser';
import { type ComboState, newComboState, nextCombo } from './comboPitch';

/**
 * Procedural sound engine — the game's only audio. Soft synth bells/marimba,
 * synthesised live with the Web Audio API (no asset files, matching the
 * procedural-textures rule). A singleton, mirroring the one shared scene/bus:
 * each zone calls into it at its own local hook point.
 *
 * Disabled-until-init and self-silencing: every cue is a no-op until `init`
 * finds a Web Audio context, so importing it never makes noise and HTML5/NoAudio
 * fallbacks degrade cleanly.
 */

/** Overall output level — one knob for the whole mix (0–1). */
const MASTER_GAIN = 0.6;
/** Merges/multiplications closer together than this (ms) climb in pitch. */
const COMBO_WINDOW_MS = 1500;

interface ToneOpts {
  freq: number;
  type?: OscillatorType;
  /** Fade-in, seconds. */
  attack?: number;
  /** Fade-out after the attack, seconds. */
  decay?: number;
  /** Peak gain before the master bus. */
  gain?: number;
  /** If set, glide the pitch to this frequency over the tone's life. */
  glideTo?: number;
  /** Start offset from now, seconds (for arpeggios). */
  delay?: number;
}

class SfxEngine {
  private ctx?: AudioContext;
  private master?: GainNode;
  private muted = false;

  // Independent chains: a merge streak and a multiply streak escalate separately.
  private readonly mergeCombo: ComboState = newComboState();
  private readonly multiplyCombo: ComboState = newComboState();

  /**
   * Reuse Phaser's AudioContext so its mobile autoplay-unlock gesture covers us.
   * Safe to call repeatedly; only the first wins.
   */
  init(scene: Phaser.Scene): void {
    if (this.ctx) return;
    const ctx = (scene.sound as unknown as { context?: AudioContext }).context;
    if (!ctx) return; // HTML5 / NoAudio manager — stay a no-op
    this.ctx = ctx;
    this.master = ctx.createGain();
    this.master.gain.value = MASTER_GAIN;
    this.master.connect(ctx.destination);
  }

  setMuted(muted: boolean): void {
    this.muted = muted;
    if (this.master) this.master.gain.value = muted ? 0 : MASTER_GAIN;
  }

  toggleMute(): void {
    this.setMuted(!this.muted);
  }

  // --- Cues, tuned by relevance (rarer/more important = louder) -------------

  /** Zone A: a ball is released. Soft low thunk. */
  drop(): void {
    this.tone({ freq: 180, type: 'triangle', attack: 0.004, decay: 0.12, gain: 0.18, glideTo: 120 });
  }

  /** Zone A: two balls merge. Bell ping that climbs through a fast chain. */
  merge(): void {
    const mult = this.combo(this.mergeCombo);
    if (mult === undefined) return;
    const base = 523.25; // C5
    this.tone({ freq: base * mult, type: 'sine', attack: 0.004, decay: 0.22, gain: 0.26 });
    this.tone({ freq: base * mult * 2, type: 'sine', attack: 0.004, decay: 0.12, gain: 0.08 });
  }

  /** Zone C: a ball is sucked through the trap-door. Short descending whoosh. */
  transition(): void {
    this.tone({ freq: 520, type: 'triangle', attack: 0.004, decay: 0.2, gain: 0.22, glideTo: 180 });
  }

  /** Zone B: a ball splits into `n` copies. Bright pluck, same chain rule as merge. */
  multiply(n: number): void {
    if (n <= 1) return;
    const mult = this.combo(this.multiplyCombo);
    if (mult === undefined) return;
    const base = 392.0; // G4
    this.tone({ freq: base * mult, type: 'triangle', attack: 0.003, decay: 0.16, gain: 0.2 });
  }

  /** Zone B: a ball drains. Quiet coin tick — happens often, must not dominate. */
  collect(value: number): void {
    const semis = Math.min(12, Math.log2(Math.max(1, value))); // tint up with value
    this.tone({ freq: 880 * 2 ** (semis / 12), type: 'sine', attack: 0.002, decay: 0.07, gain: 0.09 });
  }

  /** Zone A: one ball buffer slot arrives during a score-bar cash-in refill. Ascending
   *  blip, `index`-th in the refill sequence (0-based) — climbs a semitone per step like
   *  the merge/multiply combo chains, but driven directly by the caller's own counter since
   *  buffer ticks already run on a fixed cadence rather than player-timed hits. */
  bufferTick(index: number): void {
    const base = 784.0; // G5
    const mult = 2 ** (Math.min(index, 8) / 12);
    this.tone({ freq: base * mult, type: 'sine', attack: 0.003, decay: 0.1, gain: 0.16 });
  }

  /** Score bar filled. Rising major-triad arpeggio — the loudest, rarest cue. */
  goal(): void {
    const notes = [659.25, 830.61, 987.77]; // E5, G#5, B5
    notes.forEach((freq, i) => {
      this.tone({ freq, type: 'sine', attack: 0.005, decay: 0.28, gain: 0.5, delay: i * 0.09 });
    });
  }

  // --- Internals -----------------------------------------------------------

  /** Advance a combo chain; undefined when audio is off (skip the cue). */
  private combo(state: ComboState): number | undefined {
    if (!this.ctx) return undefined;
    return nextCombo(state, this.ctx.currentTime * 1000, COMBO_WINDOW_MS).mult;
  }

  /** One short voice: osc → gain (exp attack/decay) → master, auto-stopped. */
  private tone(o: ToneOpts): void {
    const ctx = this.ctx;
    const master = this.master;
    if (!ctx || !master || this.muted) return;

    const start = ctx.currentTime + (o.delay ?? 0);
    const attack = o.attack ?? 0.005;
    const decay = o.decay ?? 0.18;
    const peak = o.gain ?? 0.2;
    const end = start + attack + decay;

    const osc = ctx.createOscillator();
    osc.type = o.type ?? 'sine';
    osc.frequency.setValueAtTime(o.freq, start);
    if (o.glideTo !== undefined) {
      osc.frequency.exponentialRampToValueAtTime(Math.max(1, o.glideTo), end);
    }

    const g = ctx.createGain();
    g.gain.setValueAtTime(0.0001, start); // exp ramps need a positive start
    g.gain.exponentialRampToValueAtTime(peak, start + attack);
    g.gain.exponentialRampToValueAtTime(0.0001, end);

    osc.connect(g).connect(master);
    osc.start(start);
    osc.stop(end + 0.02);
  }
}

/** The shared procedural sound engine. */
export const Sfx = new SfxEngine();

File: Assets/Game/Gameplay/Juice/Sfx.cs
using RichCoast.Core;
using UnityEngine;

namespace RichCoast.Game
{
    /// <summary>
    /// Procedural placeholder audio — synthesised <see cref="AudioClip"/>s (soft sine/triangle bells
    /// with an exponential envelope), no asset files, mirroring the web build's Web Audio engine.
    /// Cues pitch-climb through fast chains via the pure, unit-tested <see cref="ComboPitch"/>,
    /// applied as <see cref="AudioSource.pitch"/>. Replace clips with authored SFX later without
    /// touching callers.
    /// </summary>
    public sealed class Sfx : MonoBehaviour
    {
        const int SampleRate = 44100;
        static Sfx instance;

        GameFeelSO feel;
        AudioSource[] voices;
        int nextVoice;
        AudioClip drop, merge, mergeHigh, tick, goal, gameOver, transition, multiply, collect;
        readonly ComboState mergeCombo = new ComboState();
        readonly ComboState multiplyCombo = new ComboState();
        bool muted;

        public static Sfx Instance => instance;

        public static Sfx Create(GameFeelSO feel)
        {
            if (instance != null) return instance;
            var go = new GameObject("Sfx");
            DontDestroyOnLoad(go);
            instance = go.AddComponent<Sfx>();
            instance.feel = feel;
            instance.Build();
            return instance;
        }

        void Build()
        {
            voices = new AudioSource[8];
            for (int i = 0; i < voices.Length; i++)
            {
                var src = gameObject.AddComponent<AudioSource>();
                src.playOnAwake = false;
                src.spatialBlend = 0f;
                voices[i] = src;
            }
            drop = Synth("drop", 180f, 120f, Wave.Triangle, 0.004f, 0.12f, 0.18f);
            merge = Layer("merge", Synth("m1", 523.25f, 523.25f, Wave.Sine, 0.004f, 0.22f, 0.26f), Synth("m2", 1046.5f, 1046.5f, Wave.Sine, 0.004f, 0.12f, 0.08f));
            mergeHigh = Synth("mergeHigh", 659.25f, 659.25f, Wave.Sine, 0.004f, 0.3f, 0.3f);
            tick = Synth("tick", 784f, 784f, Wave.Sine, 0.003f, 0.1f, 0.16f);
            goal = Arpeggio("goal", new[] { 659.25f, 830.61f, 987.77f }, 0.09f, 0.28f, 0.5f);
            gameOver = Layer("gameOver", Synth("g1", 220f, 110f, Wave.Triangle, 0.01f, 0.6f, 0.3f), Synth("g2", 164.8f, 82.4f, Wave.Sine, 0.01f, 0.7f, 0.25f));
            transition = Synth("transition", 520f, 180f, Wave.Triangle, 0.004f, 0.2f, 0.22f);
            multiply = Synth("multiply", 392f, 392f, Wave.Triangle, 0.003f, 0.16f, 0.2f);
            collect = Synth("collect", 880f, 880f, Wave.Sine, 0.002f, 0.07f, 0.09f);
        }

        /// <summary>Zone C: the trap-door sucks a ball through. Downward whoosh.</summary>
        public void Transition() => Play(transition, 1f);

        /// <summary>Zone B: a ball splits into <paramref name="n"/> copies. Bright pluck that climbs through a fast chain.</summary>
        public void Multiply(int n)
        {
            if (n <= 1) return;
            double mult = ComboPitch.Next(multiplyCombo, Time.unscaledTime * 1000.0, feel.comboWindowMs, feel.comboMaxStep).Mult;
            Play(multiply, (float)mult);
        }

        /// <summary>Zone B: a ball drains. Quiet coin tick, tinted up with value — happens often, must not dominate.</summary>
        public void Collect(double value)
        {
            double semis = System.Math.Min(12, System.Math.Log(System.Math.Max(1, value), 2));
            Play(collect, Mathf.Pow(2f, (float)semis / 12f));
        }

        public void SetMuted(bool on) => muted = on;
        public void ToggleMute() => muted = !muted;

        /// <summary>Zone A: a ball is released. Soft low thunk.</summary>
        public void Drop() => Play(drop, 1f);

        /// <summary>Zone A: two balls merge. Bell ping that climbs through a fast chain.</summary>
        public void Merge(int tier)
        {
            double mult = ComboPitch.Next(mergeCombo, Time.unscaledTime * 1000.0, feel.comboWindowMs, feel.comboMaxStep).Mult;
            Play(tier >= feel.heavyHapticFromTier ? mergeHigh : merge, (float)mult);
        }

        /// <summary>One buffer slot arrives during a refill: ascending blip, index-th in the sequence.</summary>
        public void BufferTick(int index) => Play(tick, Mathf.Pow(2f, Mathf.Min(index, 8) / 12f));

        /// <summary>Score bar filled / round banked. The loudest, rarest cue.</summary>
        public void Goal() => Play(goal, 1f);

        public void GameOver() => Play(gameOver, 1f);

        void Play(AudioClip clip, float pitch)
        {
            if (muted || clip == null) return;
            var src = voices[nextVoice];
            nextVoice = (nextVoice + 1) % voices.Length;
            src.pitch = pitch;
            src.volume = feel.masterVolume;
            src.clip = clip;
            src.Play();
        }

        // --- synthesis ------------------------------------------------------------------

        enum Wave { Sine, Triangle }

        /// <summary>One voice: osc (freq gliding to glideTo) × exp attack/decay envelope × gain.</summary>
        static AudioClip Synth(string name, float freq, float glideTo, Wave wave, float attack, float decay, float gain)
        {
            float length = attack + decay + 0.02f;
            int n = Mathf.CeilToInt(length * SampleRate);
            var data = new float[n];
            double phase = 0;
            for (int i = 0; i < n; i++)
            {
                float t = i / (float)SampleRate;
                float u = Mathf.Clamp01(t / (attack + decay));
                float f = freq * Mathf.Pow(glideTo / freq, u); // exponential glide
                phase += f / SampleRate;
                float env = t < attack ? Mathf.Pow(t / attack, 0.5f) : Mathf.Exp(-(t - attack) / decay * 5f);
                float s = wave == Wave.Sine ? Mathf.Sin((float)(phase * Mathf.PI * 2)) : Triangle((float)phase);
                data[i] = s * env * gain;
            }
            var clip = AudioClip.Create(name, n, 1, SampleRate, false);
            clip.SetData(data, 0);
            return clip;
        }

        static float Triangle(float phase)
        {
            float p = phase - Mathf.Floor(phase);
            return 4f * Mathf.Abs(p - 0.5f) - 1f;
        }

        static AudioClip Layer(string name, params AudioClip[] clips)
        {
            int n = 0;
            foreach (var c in clips) n = Mathf.Max(n, c.samples);
            var mix = new float[n];
            foreach (var c in clips)
            {
                var d = new float[c.samples];
                c.GetData(d, 0);
                for (int i = 0; i < d.Length; i++) mix[i] += d[i];
            }
            var clip = AudioClip.Create(name, n, 1, SampleRate, false);
            clip.SetData(mix, 0);
            return clip;
        }

        static AudioClip Arpeggio(string name, float[] notes, float gapS, float decay, float gain)
        {
            int n = Mathf.CeilToInt((gapS * notes.Length + decay + 0.05f) * SampleRate);
            var mix = new float[n];
            for (int k = 0; k < notes.Length; k++)
            {
                var voice = Synth("v", notes[k], notes[k], Wave.Sine, 0.005f, decay, gain);
                var d = new float[voice.samples];
                voice.GetData(d, 0);
                int offset = Mathf.RoundToInt(k * gapS * SampleRate);
                for (int i = 0; i < d.Length && offset + i < n; i++) mix[offset + i] += d[i];
            }
            var clip = AudioClip.Create(name, n, 1, SampleRate, false);
            clip.SetData(mix, 0);
            return clip;
        }
    }
}
